Kitodo-API/src/main/java/org/kitodo/api/dataformat/SortedList.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * (c) Kitodo. Key to digital objects e. V. <contact@kitodo.org>
+ *
+ * This file is part of the Kitodo project.
+ *
+ * It is licensed under GNU General Public License version 3 or later.
+ *
+ * For the full copyright and license information, please read the
+ * GPL3-License.txt file that was distributed with this source code.
+ */
+
+package org.kitodo.api.dataformat;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.function.Function;
+
+/**
+ * A list that is aware of and enforces the order of its members prescribed by
+ * the ordinal number of the media units.
+ */
+class SortedList<T> extends ArrayList<T> {
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * Possibility to get an order number for T.
+     */
+    private Function<T, Integer> orderGetter;
+
+    /**
+     * Constructor for an order-aware list.
+     * 
+     * @param orderGetter
+     *            possibility to get an order number for T
+     */
+    public SortedList(Function<T, Integer> orderGetter) {
+        this.orderGetter = orderGetter;
+    }
+
+    /**
+     * Sorts the list by the media unit’s order value and can handle duplicate
+     * occurrences of the same number (unlike a TreeMap). The order of list
+     * items is changed only when necessary. The order of list members with the
+     * same ordinal number is not affected.
+     */
+    private void ensureOrder() {
+        Collections.sort(this, (one, another) -> orderGetter.apply(one).compareTo(orderGetter.apply(another)));
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        ensureOrder();
```

Comment:
....and then you don't need to call it on every other method

Reply:
At first we should ask ourself if this list needs to be sorted all the time or only in specific moment of usage like iteration over it. I have a feeling that it is a bit overkill to sort elements before adding new element, sort elements before iterator and so on. I assume that only moment when order really matters is iterating over it. Good solution for it is override compare method and next sort list right before you want to usage it.